### PR TITLE
Use auto_increment IDs, remove mapping tables, plus more

### PR DESCRIPTION
**Pull request by MatthewBemis**
For GAWB-515

-Split ATTRIBUTE table into three new tables: SUBMISSION_ATTRIBUTE, ENTITY_ATTRIBUTE, WORKSPACE_ATTRIBUTE
-Remove ENTITY_ATTRIBUTE and WORKSPACE_ATTRIBUTE mapping tables
-Switch all attribute tables and ENTITY table to use auto-incrementing IDs instead of UUIDs
-Improves performance for any query that joined on the attributes tables, as well as significant gains in clone workspace, import TSV, and copy entities (cuts number of queries by about ~50%)

For reviewer: old schema:

```
ATTRIBUTE
id <<PK>> | name | value_string | value_number | value_boolean | value_entity_ref <<FK>> | list_index

ENTITY
id <<PK>> | name | entity_type | workspace_id <<FK>>

ENTITY_ATTRIBUTE
attribute_id <<PK>> <<FK>> | entity_id <<FK>>

WORKSPACE_ATTRIBUTE
attribute_id <<PK>> <<FK>> | workspace_id <<FK>>

Submission validation values held a direct link to an ATTRIBUTE id, no mapping table needed.
```

new schema:

```
ENTITY
id <<PK>> | name | entity_type | workspace_id <<FK>>

ENTITY_ATTRIBUTE
id <<PK>> | name | value_string | value_number | value_boolean | value_entity_ref <<FK>> | list_index | entity_id <<FK>>

WORKSPACE_ATTRIBUTE
id <<PK>> | name | value_string | value_number | value_boolean | value_entity_ref <<FK>> | list_index | workspace_id <<FK>>

SUBMISSION_ATTRIBUTE
id <<PK>> | name | value_string | value_number | value_boolean | value_entity_ref <<FK>> | list_index | validation_id <<FK>>
```
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160428_entity_attrs_autoinc.xml [hunk 1]
```diff
@@ -0,0 +1,335 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog logicalFilePath="dummy" xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
```
Comment (MatthewBemis) [inline src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160428_entity_attrs_autoinc.xml]: Ignore this for now. It's not 100% complete.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala [hunk 2]
```diff
@@ -13,34 +13,377 @@ import spray.http.StatusCodes
 /**
  * Created by dvoet on 2/4/16.
  */
-case class AttributeRecord(id: UUID,
+case class EntityAttributeRecord(id: Long,
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala:27]: There's a lot of similar/repeated code in this AttributeComponent class. We have three types of attributes, each one differing slightly (entity attributes have a Long id to an entity, workspace attributes have a UUID to a workspace, and submission attributes have no foreign key to a parent)

Additionally, slick doesn't seem to provide a way to reuse code for different tables, so even if the data signatures matched up, we'd possibly still be stuck with duplicate code

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala [hunk 3]
```diff
@@ -285,57 +269,44 @@ trait EntityComponent {
...
-      }.grouped(batchSize).toSeq
-
-      DBIO.seq(insertBatches.map { insertBatch =>
-        val prefix = sql"insert into ENTITY (id, name, entity_type, workspace_id) values "
-        val suffix = insertBatch.reduce { (a, b) =>
-          concatSqlActionsWithDelim(a, b, sql", ")
-        }
-        concatSqlActions(prefix, suffix).as[Int]
-      }:_*)
+    def batchInsertEntities(workspaceContext: SlickWorkspaceContext, entities: Seq[EntityRecord]): ReadWriteAction[Seq[EntityRecord]] = {
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala:271]: The new batchInsertEntities combined with the new schema allows us to insert batches of entities, select back their IDs (batching the selects as well), and then use those ids to insert attributes into the ENTITY_ATTRIBUTE table. We no longer need the attribute ids back, which was about 75% of the battle.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 4]
```diff
@@ -583,7 +581,8 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: SlickDataSo
           if (!errorReports.isEmpty) {
             DBIO.failed(new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Some entities could not be updated.", errorReports)))
           } else {
-            dataAccess.entityQuery.save(workspaceContext, updateTrials.collect { case (entityUpdate, Success(entity)) => entity } )
+            val t = updateTrials.collect { case (entityUpdate, Success(entity)) => entity }
+            dataAccess.entityQuery.save(workspaceContext, t )
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:585]: dunno why anything in this class changed, I'll put it back

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala [hunk 5]
```diff
@@ -14,7 +14,7 @@ import org.joda.time.DateTime
 class EntityComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers {
   import driver.api._
 
-  ignore should "*REENABLE WITH MYSQL UNIT TESTS* crud entities" in withEmptyTestDatabase {
```
Comment (MatthewBemis) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponentSpec.scala:17]: the delete from a join raw sql that caused all of these unit tests to break is no longer needed

**Comment by MatthewBemis** [comment]
~~Note to self: I need more tests for these new attribute tables~~

they all use the same code now.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala [hunk 6]
```diff
@@ -13,34 +13,377 @@ import spray.http.StatusCodes
...
                            valueNumber: Option[Double],
                            valueBoolean: Option[Boolean],
-                           valueEntityRef: Option[UUID],
+                           valueEntityRef: Option[Long],
                            listIndex: Option[Int])
 
+case class WorkspaceAttributeRecord(id: Long,
+                                    workspaceId: UUID,
+                                    name: String,
+                                    valueString: Option[String],
```
Comment (dmohs) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala:39]: You might consider:

``` scala
type ValueString = Option[String]
```

etc. to get some reuse here. Probably not worth doing but wanted to offer the suggestion anyway.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala [hunk 7]
```diff
@@ -13,34 +13,377 @@ import spray.http.StatusCodes
...
+
+    def entityRef = foreignKey("FK_ENT_ATTRIBUTE_ENTITY_REF", valueEntityRef, entityQuery)(_.id.?)
+    def parentEntity = foreignKey("FK_ATTRIBUTE_PARENT_ENTITY", entityId, entityQuery)(_.id)
+  }
+
+  class WorkspaceAttributeTable(tag: Tag) extends Table[WorkspaceAttributeRecord](tag, "WORKSPACE_ATTRIBUTE") {
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def workspaceId = column[UUID]("workspace_id")
     def name = column[String]("name")
     def valueString = column[Option[String]]("value_string")
```
Comment (dmohs) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala]: Actually, if this type has to match the above then it probably is worth doing.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala [hunk 8]
```diff
@@ -13,34 +13,377 @@ import spray.http.StatusCodes
...
+    /**
+     * Insert an attribute into the database. This will be multiple inserts for a list and will lookup
+     * referenced entities.
+     *
+     * @param name
+     * @param attribute
+     * @param workspaceId used only for AttributeEntityReferences (or lists of them) to resolve the reference within the workspace
+     * @return a sequence of write actions the resulting value being the attribute id inserted
+     */
+    def insertAttributeRecords(entityId: Long, name: String, attribute: Attribute, workspaceId: UUID): Seq[ReadWriteAction[Int]] = {
```
Comment (dmohs) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala]: The similarity between these methods and those similarly named below worries me a great deal from a code maintenance perspective. Can you factor out the common functionality? I can offer some suggestions if that would help.

Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala]: I have given a solution in https://github.com/broadinstitute/rawls/pull/399

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala [hunk 9]
```diff
@@ -166,23 +150,22 @@ trait EntityComponent {
         _ <- deleteEntityAttributes(preExistingEntityRecs)
         savingEntityRecs <- insertNewEntities(workspaceContext, entities, preExistingEntityRecs).map(_ ++ preExistingEntityRecs)
         referencedAndSavingEntityRecs <- lookupNotYetLoadedReferences(workspaceContext, entities, savingEntityRecs).map(_ ++ savingEntityRecs)
-        _ <- insertAttributes(entities, referencedAndSavingEntityRecs)
+        _ <- insertAttributes(workspaceContext, entities, referencedAndSavingEntityRecs)
       } yield entities
     }
 
-    private def insertAttributes(entities: Traversable[Entity], entityRecs: Traversable[EntityRecord]) = {
+    private def insertAttributes(workspaceContext: SlickWorkspaceContext, entities: Traversable[Entity], entityRecs: Traversable[EntityRecord]) = {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala]: why did you add workspaceContext here? It does not seem to be used.

Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala]: ah, it used to be used in an older commit of this branch but I forgot to remove it

**Comment by MatthewBemis** [comment]
@dmohs I forgot to mention yesterday, but the liquibase is ready.

Also notable changes coming in this PR soon: 

-Merge Doug's solution for making attribute code generic
-Make submission attributes hold an ID to their parent submission_validations so that they'll be of the same form as the other attributes and can use the generic solution

**Comment by dmohs** [comment]
👍 Much cleaner. Awesome.

File: src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160428_entity_attrs_autoinc.xml [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+
+    <!--Remove some foreign keys-->
+
+    <changeSet logicalFilePath="dummy" author="mbemis" id="entity-attrs-autoinc-1">
+        <sql stripComments="true">
+            alter table ATTRIBUTE drop foreign key ATTRIBUTE_ibfk_1;
```
Comment (dvoet) [inline src/main/resources/org/broadinstitute/dsde/rawls/liquibase/changesets/20160428_entity_attrs_autoinc.xml]: I am concerned that these auto generated foreign key names might be different on prod, double check them.

Name the foreign keys you create below to avoid this in the future.

**Comment by MatthewBemis** [comment]
Doge gave a verbal :+1:
